### PR TITLE
Adding lower bound to pnetcdf conflict in Trilinos

### PR DESCRIPTION
**Pull request by jrood-nrel**
Without this lower bound, the wrong conflicts occur when using non-numbered versions of Trilinos.

### CODE REVIEW

File: var/spack/repos/builtin/packages/trilinos/package.py [hunk 1]
```diff
@@ -173,7 +173,7 @@ class Trilinos(CMakePackage):
     # deprecated SuperLUDist v3.3 together with an Amesos patch is working.
     conflicts('+superlu-dist', when='@11.4.1:11.14.3')
     # PnetCDF was only added after v12.10.1
-    conflicts('+pnetcdf', when='@:12.10.1')
+    conflicts('+pnetcdf', when='@11.4.1:12.10.1')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/trilinos/package.py]: Every time someone adds an older version of `trilinos`, this number will need to be changed. If you're only trying to exclude `master`, why not set the lower bound to 0?
```
$ spack python
Spack version 0.10.0
Python 2.7.13, Darwin x86_64
>>> from spack.version import *
>>> Version('master') < Version('0')
True
```

**Comment by adamjstewart** [comment]
Ping @davydden @citibeth @tgamblin. Not to reopen a can of worms, but we really need to convince Spack that `master` is indeed greater than all numeric releases, and less than `develop`. We already make an exception for `develop`, I don't think it would be much work to make an exception for `master` too. I'm tired of telling people to rename `master` to `develop` just because Spack is broken. Could we just say that all non-numeric versions are greater than numeric versions?

**Comment by davydden** [comment]
>  I don't think it would be much work to make an exception for master too. 

The work is done long time ago https://github.com/LLNL/spack/pull/1983 and the extension to `master` is a simple change of this line https://github.com/davydden/spack/blob/a13a571cfdb111afdf3671ebda9672604c2000ba/lib/spack/spack/version.py#L60

> Could we just say that all non-numeric versions are greater than numeric versions?

i would not do this as there are also fancy cases of `xsd-0.1` and alike. You would also need to define what _non-numeric_ is. From my perspective (and that's how the algorithm works right now), there is no need in distinguishing between numeric and non-numeric versions (!), one only needs to sort numbers and strings. That is, `0.1.abc` is broken into `0`, `1` and `abc` and then being compared to, say, `0.2` -> `0`, `2`. 

What I am saying is, https://github.com/LLNL/spack/pull/1983 make extra comparison rules for the pre-chosen set of names to make `0.2.develop` > `0.2.1`.

**Comment by adamjstewart** [comment]
I think you actually backed this change out of #1983, but it wouldn't be hard to re-add it.

> there are also fancy cases of `xsdk-0.1` and alike.

You mean `0.1-xsdk`? Problem solved 😉 

> From my perspective (and that's how the algorithm works right now), there is no need in distinguishing between numeric and non-numeric versions

That's not actually true. Spack needs to sort versions, but it also needs to choose the preferred version, which is currently the newest non-numeric version.

**Comment by davydden** [comment]
> choose the preferred version, which is currently the newest non-numeric version.

I don't think so, it's the newset non-develop version

**Comment by davydden** [comment]
> I think you actually backed this change out of #1983, but it wouldn't be hard to re-add it.

Exactly!

**Pull request merged**